src/main/java/kr/co/vcnc/haeinsa/HaeinsaPut.java

```diff
@@ -121,6 +127,45 @@ public void add(HaeinsaMutation newMutation) {
         }
     }
 
+    @VisibleForTesting
+    void remove(HaeinsaDelete delete) {
+        HaeinsaKeyValue searchFromKV = new HaeinsaKeyValue();
+        // Type is descending ordered in HaeinsaKeyValue.compareTo()
+        searchFromKV.setType(Maximum);
+        searchFromKV.setRow(row);
+
+        HaeinsaKeyValue searchToKV = new HaeinsaKeyValue();
+        searchToKV.setType(Minimum);
+        searchToKV.setRow(row);
+
+        for (Map.Entry<byte[], NavigableSet<HaeinsaKeyValue>> deleteEntry : delete.getFamilyMap().entrySet()) {
+            byte[] family = deleteEntry.getKey();
+            NavigableSet<HaeinsaKeyValue> putKeyValues = familyMap.get(family);
+            if (putKeyValues == null) {
+                continue;
+            }
+            searchFromKV.setFamily(family);
+            searchToKV.setFamily(family);
+            for (HaeinsaKeyValue deleteKeyValue : deleteEntry.getValue()) {
+                List<HaeinsaKeyValue> removingList;
```

Comment:
How about eliminate the list allocation by using NavigableSet directly for iteration?